### ISSUE TITLE
Make RAM (and CPU) - usage visually intuitive

### ISSUE THREAD

**Issue by GammaSQ** [QubesOS/qubes-issues#4531]
### Qubes OS version: 4.0
I miss the RAM and CPU usage I could check in qubes-vm-manager in Qubes 3.X

Right now, I use `xentop` if my fan speeds up to figure out which VM is pulling more load than expected, but this isn't as straight forward as having a list of running qubes with their usage. Especially glancing at usage to, sometimes, identify VMs pulling more than expected without raising the fan-speed (that 10%-VM that should be at 1%, because some process didn't quit gracefully) is just not possible without an existing, easy to check, hands off usage-list.

It was briefly mentioned on the mailinglist that RAM and CPU usage could be added to qvm-ls. Though that would be a first step, at least for me the visual cues qubes manager provided would be invaluable. Colors are easier glanced at than names, bars are faster parsed than numbers, ALT+tabtabtab is faster than typing e.g. "qvm-ls --show-me-my-usage".

Since the plan is to move away from qubes manager, I'm not sure where this would be best implemented. My initial reaction would be it makes no sense anywhere but the Qubes Manager, but I'm no UX-designer.

**Comment by marmarek** [comment]
See domain widget - currently it list RAM usage, but we plan to add CPU usage there too (#4401).

**Comment by GammaSQ** [comment]
Oh! The gray numbers are RAM-usage? I thought they were disk-space.
They are quite hard to read. Is there a way to set the color-scheme for RAM-usage?

Even better would be to enable bars. With numbers I have to use my brain to calculate how much I'm using where. Bars are a lot faster to understand visually. I enjoyed having both in the qubes VM manager, but I guess there would be too little space with the widget.

**Comment by marmarek** [comment]
> Oh! The gray numbers are RAM-usage? I thought they were disk-space.
> They are quite hard to read. Is there a way to set the color-scheme for RAM-usage?

Hmm, they are black on my system...
It may be affected by xfce color settings (see system settings).

**Comment by GammaSQ** [comment]
> Hmm, they are black on my system...
> It may be affected by xfce color settings (see system settings).
Indeed they were!
I'll compose a new issue for the QubesDomains-widget.
Leaving this one open as wishing for bars instead of numbers.

**Comment by airelemental** [comment]
I just updated, the CPU usage is really nice.

Would it be possible to move the CPU percent usage to the left of (or inside of) the cpu bar, and drop the "CPU" label? This would be much more space-efficient, right now about 40% of the space is unused. Everything would also line up, rather than alternating CPU % and CPU bar, for easier comparison.

**Comment by marmarta** [comment]
@airelemental the problem with omitting the CPU label is that it's really not obvious this is, in fact, the CPU usage. If you don't expect CPU (or don't know to expect it), it's not obvious this is not in some way related to RAM MB to the side. I had even shown that to a very technical user (without telling them it's the CPU percentage) and they spend some time being seriously puzzled over why the same amount of MB corresponds to different percentages; I'm afraid this will be even more confusing to most people.

I did not have any success with placing the label inside of the bar (and maintaining readability) or with placing it to the left and maintaining clarity and readability. The current solution is the one I found the most readable; if you can try to get Gtk to have it look better, any pull requests are welcome :)

**Comment by airelemental** [comment]
> @airelemental the problem with omitting the CPU label is that it's really not obvious this is, in fact, the CPU usage. If you don't expect CPU (or don't know to expect it), it's not obvious this is not in some way related to RAM MB to the side. I had even shown that to a very technical user (without telling them it's the CPU percentage) and they spend some time being seriously puzzled over why the same amount of MB corresponds to different percentages; I'm afraid this will be even more confusing to most people.

I think the main tell is seeing that CPU % and memory usage behave in their characteristic ways - CPU meters rapidly fluctuate, will be high when the system feels busy, whereas the memory usages are mostly static. Although as you say some users may not get it on day one, they'll eventually learn from experience.

IMHO fixed labels may be useful once but uselessly waste space, forever. For example, all the xfce4 system monitor widgets have an option to turn off labels, Though I concede that space-efficiency is less acute since the qubes one only displays after a click, rather than the always-displayed xfce4 system widgets.

> if you can try to get Gtk to have it look better, any pull requests are welcome :)

https://github.com/airelemental/qubes-desktop-linux-manager/commit/23b37bfaf5812d4914d521087e210e53cd920bff

I'll clean it up and make a pull request, if it looks acceptable?

**Comment by marmarta** [comment]
My biggest problem with your solution is that when you get double-digit CPU usage, the RAM column moves to the side: 
![moved](https://user-images.githubusercontent.com/7242298/59132458-86f6a200-8975-11e9-8233-cd518685f1ea.png)
(I have not managed to wrangle that bit of Gtk down, I admit)

**Comment by marmarta** [comment]
Also, by @marmarek 's request, I'm adding a screenshot of my version:
![domains2](https://user-images.githubusercontent.com/7242298/59132606-f7052800-8975-11e9-861d-52b5367f2bde.png)
(with less whitespace, I don't have a version running with the less whitespace variant)

**Comment by airelemental** [comment]
I'm new to gtk3, I'll do some reading and get that fixed.

**Comment by airelemental** [comment]
An explicit cpu_label.set_width_chars(4) fixes it for double-digit percentage, but the issue still occurs in the really rare instance of triple-digit (100%) cpu percentage. I don't want to increase the width_chars and waste space all the time for just this rare occurrence though.

https://github.com/airelemental/qubes-desktop-linux-manager/commit/052ed1909bc1743a78f6ab2d906071ebe62a4588

**Comment by marmarta** [comment]
Depending on the font and xfce style I choose, set_width_chars(4) works or not; likewise, with high dpi adjustments and some fonts I can even get set_width_chars(5) to misbehave. And given how terrible RAM dancing around looks, I'm not much of a fan of risking it happen.

**Comment by marmarek** [comment]
I don't have a problem with CPU label + number over the bar. Given its gray-ish color, it's easy enough to compare bars which have strong color. IMO it looks nice this way and it is clear that the CPU% is presented in two forms.
My only problem with the current solution is that each row got a bit thicker. I know @marmarta tried to put label + number on the bar and failed (I've heard a lot of "opinions about Gtk3" in the process). But maybe bars could be a bit thinner and/or label smaller to maintain consistent row height?

**Comment by airelemental** [comment]
> Depending on the font and xfce style I choose, set_width_chars(4) works or not; likewise, with high dpi adjustments and some fonts I can even get set_width_chars(5) to misbehave. And given how terrible RAM dancing around looks, I'm not much of a fan of risking it happen.

It seems like with variable-width fonts, set_width_chars can only be approximate, so a safety margin is needed. The "%" character in e.g. liberation-sans renders almost twice as wide as 2 digits, so that's why I thought 4 would be sufficient for 2 digits and "%".

What about set_width_chars(6) then? It should accommodate even 3 characters that are all double-width.

**Comment by airelemental** [comment]
What if the bars were dropped entirely, leaving only the CPU percent numbers?
* The bars don't convey much information for the space they take
* They are redundant with the CPU percent numbers
* At any given time, most of the bars are 0 anyways

![menu](https://user-images.githubusercontent.com/40210726/59140405-3c345480-898c-11e9-9103-b1e6746f8174.png)

Then later, the CPU usage units can be changed to number of processor cores, e.g. 151% (like in xentop), which doesn't lend itself to being represented as a bar anyway. But would be more informative.

**Comment by marmarta** [comment]
I definitely would like some description with the percentage - it's not good to have pure, undescribed numbers anywhere. I know we know what they mean, but I would hate a situation in which someone keeps looking at them for months on end being slightly confused (and with no easy way of fixing that confusion). Maybe a tooltip could work, but on the other hand, you'd have to guess that a drop-down menu has tooltips, which is also not obvious for many people. 
@andrewdavidwong , @marmarek, anyone - I think we need more opinions here, because it looks to me that we have two people with opposed ideas on what looks nice and is useful here :)

**Comment by andrewdavidwong** [comment]
In both cases, a table is being used to display information, so why not have a row of headings for the columns in that table?

| Qube | RAM | CPU |
| ---- | --- | --- |
| vault | 400 MB | 12% |
| personal | 650 MB | 21% |
| web | 800 MB | 32% |
| work | 420 MB | 14% |

You could also include visual bars in either the RAM or CPU cells (or both).

**Comment by airelemental** [comment]
https://github.com/airelemental/qubes-desktop-linux-manager/commit/cdf0ba28ffb0af4fb2bb07a0a706581adbb1d113

- adds headers
- suppresses all 0% CPU entries, to make non-0% CPU usage more visible.

![menu2](https://user-images.githubusercontent.com/40210726/59330238-75e4c280-8ce0-11e9-8388-08a47bd7a081.png)

**Comment by marmarek** [comment]
I like it! Care to create a PR?

**Comment by marmarta** [comment]
It looks great!

**Comment by Le1b1** [comment]
I'd prefer 'RAM' over 'Memory' to avoid confusion.

**Comment by airelemental** [comment]
https://github.com/airelemental/qubes-desktop-linux-manager/commit/766f2d9a64a81ddd9999dbf4b74e2a7ce323971b

- remove empty submenu from header
- "RAM" instead of "memory"

![menu3](https://user-images.githubusercontent.com/40210726/59377370-9bee7f00-8d41-11e9-9374-50a1b5fff80c.png)

> Care to create a PR?

Will do, I think I want to run it a day on my system to make sure no bugs.

So far, I've noticed that the cpu usage will exceed 100% when shutting down VMs, but I think it affects the current-testing version of the widget too.

**Comment by marmarek** [comment]
Merged to master.

**Comment by airelemental** [comment]
Add dom0? Its memory and CPU usage is large relative to a typical qube so I think they should be visible in any comprehensive view of the system. Also the stats for dom0 are tracked anyway, but no menu item is created due to not seeing dom0's domain-started event.

![menu-dom0](https://user-images.githubusercontent.com/40210726/60328937-d6c9f700-997e-11e9-8462-61145420450d.png)

Submenu and tooltip for dom0 are disabled since they don't apply.

**Comment by andrewdavidwong** [comment]
If dom0 is added, I'd vote for always keeping it at the top of the list.

**Comment by brendanhoar** [comment]
> If dom0 is added, I'd vote for always keeping it at the top of the list.

Agreed.

Two sub-issues to report:

1. Running R4.0x current-testing, I'm seeing CPU jump into the xx,xxx% range during shutdown, including what appears to be a ghost entry for a shutdown that finished ok. I kept choosing shutdown over and over and nothing was happening. I chose "Run Terminal" and received a popup notice that the VM was starting! So, I guess it wasn't really running. Next time I'll check xentop to be sure.

2. The Q icon seems non-response significantly more often than it used to. Other sibling icons' drop down menus work, and the Q doesn't respond for a while, then it starts working fine and I can see the list of VMs in the drop down.

B

**Comment by airelemental** [comment]
![menu-dom0-2](https://user-images.githubusercontent.com/40210726/60489761-a511a280-9c94-11e9-87ed-6538b668cd77.png)

https://github.com/airelemental/qubes-desktop-linux-manager/commit/14569def4f5d580fe9098d0663ae3e4954131736

**Comment by marmarta** [comment]
I wonder if this non-responsiveness is some new artifact of recent rewrites, or some other problem... @brendanhoar , have you identified any scenarios such as "Q becomes unresponsive after using it to start/shut down a VM" or "after opening settings from it"? I have troubles reproducing this bug, but this may be the haphazard state of my testing system. Do you have very many running VMs? (Very many: over a dozen or something like this)

**Comment by brendanhoar** [comment]
@marmarek - not over a dozen, but I often sometimes 2-4 VMs in parallel (e.g. when I'm doing the daily multi-template parallel update run).

I can generally reproduce non-responsiveness by starting a single VM and then immediately  rapidly clicking on the Q icon. There are at least a couple of major lags most times before the Domain xxxxx is starting notification goes away, but then it starts working normally again. Sometimes a dozen or more clicks go in without a response.

Brendan

**Comment by marmarta** [comment]
@brendanhoar  that's very interesting, because I don't get this lag even when I have ~10 VMs running and start a new one. I wonder - has it always been like that? Or just recently? Are you running the current-testing version or the normal one?

**Comment by brendanhoar** [comment]
Pretty sure it’s a new issue, though with more info shown on the Q menu, maybe I am clicking the Q a lot more?

Running current-testinf

**Comment by marmarta** [comment]
So, good news is that I can reproduce this issue. Bad news: it's seems to need a partial rewrite of the qui-domains app, because it's not a slowdown on our side, it's Gtk being terrible. Good news: after a quarter bottle of rum, I managed to get it to work! There will be a PR fixing this lag incoming soon, @brendanhoar :)

**Comment by airelemental** [comment]
Sooooo, how about showing raw CPU usage, instead of normalizing?

For example, a domain with 4 cores (such as dom0 on a 4-core system) would show up to 400%, instead of 100%.

Benefits:
* Can see at a glance which domain is using the most CPU resources. Makes the CPU column entries use the same "units".
* Fixes the erratic CPU usage numbers while a domain is shutting down (apparently caused by a fluctuating denominator in the current normalization).
* Consistent with xentop
* Larger numbers convey more precision, since they are rounded down.

It's a one-line change (remove denominator from /usr/lib/python3.5/site-packages/qubes/app.py:353).

**Comment by marmarta** [comment]
I'm a bit doubtful - avoiding weird values at shutdown would be nice, but having percentages add up to more than 100% in other situations would be a bit confusing. 
@marmarek , what do you think about it?

**Comment by marmarek** [comment]
IMO weird values at shutdown doesn't warrant such change at all. On the other hand, it would make them comparable.
Anyway, I don't think having such raw value by default is a good idea. Seeing 200% CPU without explicit explanation would be confusing. As we can see in #5195, even having empty field for 0% is already confusing.
If anything, that should be an optional setting. On the widget side, not qubesd.

**Comment by marmarek** [comment]
> Fixes the erratic CPU usage numbers while a domain is shutting down (apparently caused by a fluctuating denominator in the current normalization).

I'll move normalization to actual cpu_usage calculation, instead of normalizing it at cpu_time level. It should fix this issue.

**Comment by marmarta** [comment]
I think this can be closed now - we have normalized CPU usage and RAM usage in pretty formatted numbers. ( https://github.com/QubesOS/qubes-issues/issues/4401 )
